strategies/stop-loss.js

```diff
@@ -33,20 +51,38 @@ stop.update = function(candle) {
 
 stop.log = function(candle) {
   if (this.loss !== -1) {
-    log.info('Active Stop Loss at ' + this.loss);
+    log.info('Active ' + this.stoploss.type + ' Stop Loss at ' + this.loss);
   }
 
   if (this.strategy_log) {
     this.strategy_log(candle);
   }
 }
 
+// eventually moving the stop loss price up if a trailing stop loss is used.
 // checking the stop loss price and going short if nessessary.
 // forwarding the candle to the strategy otherwise.
 stop.check = function(candle) {
+  if (this.loss === -1) {
+    if (this.strategy_check) {
+      return this.strategy_check(candle);
+    }
+
+    return;
+  }
+
+  // stop loss enabled
   var price = candle.close;
 
-  if (this.loss !== -1 && price <= this.loss) {
+  if (this.stoploss.type === 'trailing') {
+    var newLoss = price - (price * this.stoploss.percentage);
```

Comment:
My first idea was the same as this. But did some thinking during the weekend and came up with a different (not necessarily better ;) view on this. This way the absolute stop loss is growing bigger when price goes up, the other option would be to calculate a fixed amount based on the start price during init() and use that further on.

Personally I'm not sure what the correct bahavoiur is. The only experince I have with it is on Bitstamp, they use the later one with the fixed amount.

Reply:
I wasn't sure myself what the correct behavior of the trailing stop loss was and looked it up here https://www.investopedia.com/articles/trading/08/trailing-stop-loss.asp

> Purchase Price = $10
Last Price at Time of Setting Trailing Stop = $10.05
Trailing Amount = 20 cents
Immediate Effective Stop Loss Value = $9.85

>If the market price climbs to $10.97, your trailing-stop value will also rise to $10.77. If the last price now drops to $10.90, your stop value will remain intact at $10.77. If the price continues to drop, this time to $10.76, it will penetrate your stop level, immediately triggering a market order. Your order would be submitted based on a last price of $10.76. Assuming that the bid price was $10.75 at the time, the position would be closed at this point and price. The net gain would be 75 cents per share less commissions.

I've looked up the docs on bitstamp too and it seems to me that they are doing exactly the same – maintaining the stop-spread when the price rises. If i understand you correctly bitstamp also allows you to set up an amount to sell when the stop loss is triggered but this is (currently) not possible with gekko. more at #955.

> __What is Trailing Stop Order and how to use it?__
The trailing stop is more flexible than a fixed stop loss, since it automatically tracks the bitcoin’s price direction and does not have to be manually reset like the fixed stop loss.
  
> For example: Market price of bitcoin is $480 and you placed a Stop Sell Order at $450, which is in our case $30 below the current market price. Suppose price of bitcoins suddenly increases to $550. You are now in chance of “locking in” even bigger profit by manually repositioning your Stop Sell Order to higher price.

> Repositioning your Stop Order can be done automatically for you by enabling “Trailing Stop” function. In our case Trailing Stop Order will automatically reposition active Stop Order if price of bitcoin increases while maintaining spread of $30 between the bitcoins spot price and Stop Order.

https://www.bitstamp.net/article/stop-orders-and-trailing-stop-orders/

But in fact there is space for a lot of improvement here. I can imagine additional functionality like storing the initial and final stop loss price for further calculation or advice (re-)buy signals at the entry level (or percentages of it). Therefore i'd like to see different kinds of stop-loss Implementations. 
I like to do things iterative and would prefer to build a clean foundation for further extensions rather than to include additional algos in this initial pull request. It will be a solid week but i'll look at the code and improve some things if i'll find the time.

Reply:
And thanks for your ideas and comments!